### ISSUE TITLE
Allow autowrite to write even if hidden is on

### ISSUE THREAD

**Issue by cassepipe** [vim/vim#12340]
When looking for a solution for some kind of autosaving mechanism in Vim you generally end up finding about the `autowrite` option.

There's a  line about how if `hidden` is set, buffers won't be saved upon changing buffers but it is easy to overlook for a beginner that may not even even be familiar with concepts of hidden buffers, especially since they haven't had to care about it because a lot of linux distros turn hidden on when you download vim fresh off the package manager.  

Even if they read the manual they will see that `hidden` supposed to be off by default and call it a day. Now, when they are working on a project with various buffers they may  think that buffers are saved when they leave it when in fact they're not and not understand why their code is not working as expected, pulling their hair out.

Also I fail to see how it makes sense for both `hidden` and `autowrite` to be on at the same time. It just goes against common/non-seasonned expectations and it's a bad state that should not exist imho.

My proposal is that both those options be replaced by a `hide` option with 3 possible values, `prevent`, `allow`, `save`

**Comment by chrisbra** [comment]
Hm, If I understand this correctly, *autowrite*  will only be done, if Vim would otherwise close the buffer and therefore remove it entirely from memory. Now hidden does not make Vim close the buffer and therefore *autowrite* won't be considered, which I think makes sense.

Now for some kind of autosaving (which apparently is the real issue here), I would think you can use a plugin, that periodically iterates over all buffers and performs the write (this can be done independently of 'hidden' or 'autowrite' or 'confirm' options are set). I am pretty sure there are plugins out there for this already

**Comment by cassepipe** [comment]
The solution I have seen is to just set an autocommand on BufLeave rather than using a plugin but I thought an option for such a trivial thing was warranted.

Thanks for helping me understanding how it fits and works in vim. Now I realize that if hidden is not set vim, when you `:edit` a new file, vim actually *unloads* the buffer that you want to leave behind. But somehow it still appears in the buffer list ? 

It may makes sense from a technical point of view it's totally non obvious for a new user and I suspect even for a more experienced one.

Also the `:next` and :prev` and other navigating command mentioned by `autowrite` don't make much sense outside of buffer navigation to me which is only possible when `hidden` is on...

**Comment by cassepipe** [comment]
I apologized, when I did my tests I also  wanted to see if neovim was the same  and only by installing it, it squatted the `vim` command name without me realizing and I got confused and lost while trying to understand how it all worked.
I managed to untangle it and indeed that's a neovim issue but tot their credit they did document they changed the default, I missed it.
For vim the reason `hidden` was on was indeed a global config file in `/etc`

I totally understand that saving when leaving a buffer should *not* be automatic by default. I think part of my surprise was the `next` and `prev` command (and `bn` and `bp` for that matter) are implying navigation. You are temporarily leaving a buffer and you'll come back to it later, and of course by default you don't want it saved.
 It seems strange to have to decide on whether changes are to be saved upon navigation. So on the one hand I understand neovim's choice to set `hidden` on as  the default. On the other hand it's not great if you've never been exposed to what `hidden` does.

 I think I made  wrong  assumptions in that the "modern"  editors I have used before do keep files loaded upon navigation. 
Thanks for taking the time to correct me and sharing your thoughts
